source/extensions/http/custom_response/redirect_policy/redirect_policy.cc

```diff
@@ -0,0 +1,175 @@
+  RELEASE_ASSERT(downstream_headers != nullptr, "downstream_headers cannot be nullptr");
+
+  ::Envoy::Http::Utility::Url absolute_url;
+  const std::string uri(absl::StrCat(host_, path_));
+  if (!absolute_url.initialize(uri, false)) {
+    IS_ENVOY_BUG(absl::StrCat("Redirect for custom response failed: invalid location {}",
+                              absl::StrCat(host_, path_)));
+    return ::Envoy::Http::FilterHeadersStatus::Continue;
+  }
+  // Don't change the scheme from the original request
+  const bool scheme_is_http = schemeIsHttp(*downstream_headers, decoder_callbacks->connection());
+
+  // Cache original host and path
+  const std::string original_host(downstream_headers->getHostValue());
+  const std::string original_path(downstream_headers->getPathValue());
+  const bool scheme_is_set = (downstream_headers->Scheme() != nullptr);
+  // Replace the original host, scheme and path.
+  Cleanup restore_original_headers(
+      [original_host, original_path, scheme_is_set, scheme_is_http, downstream_headers]() {
+        downstream_headers->setHost(original_host);
+        downstream_headers->setPath(original_path);
+        if (scheme_is_set) {
+          downstream_headers->setScheme(scheme_is_http
+                                            ? ::Envoy::Http::Headers::get().SchemeValues.Http
+                                            : ::Envoy::Http::Headers::get().SchemeValues.Https);
+        }
+      });
+
+  downstream_headers->setScheme(absolute_url.scheme());
+  downstream_headers->setHost(absolute_url.hostAndPort());
+
+  auto path_and_query = path_;
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.http_reject_path_with_fragment")) {
+    // Envoy treats internal redirect as a new request and will reject it if URI path
+    // contains #fragment. However the Location header is allowed to have #fragment in URI path.
+    // To prevent Envoy from rejecting internal redirect, strip the #fragment from Location URI if
+    // it is present.
+    auto fragment_pos = path_.find('#');
+    path_and_query = path_.substr(0, fragment_pos);
+  }
+  downstream_headers->setPath(path_);
+
+  if (decoder_callbacks->downstreamCallbacks()) {
+    decoder_callbacks->downstreamCallbacks()->clearRouteCache();
+  }
+  // Apply header mutations before recalculating the route.
+  request_header_parser_->evaluateHeaders(*downstream_headers, decoder_callbacks->streamInfo());
+  if (modify_request_headers_action_) {
+    modify_request_headers_action_->modifyRequestHeaders(*downstream_headers,
+                                                         decoder_callbacks->streamInfo(), *this);
+  }
+  const auto route = decoder_callbacks->route();
+  // Don't allow a redirect to a non existing route.
+  if (!route) {
+    stats_.custom_response_redirect_no_route_.inc();
+    ENVOY_LOG(trace, "Redirect for custom response failed: no route found");
+    return ::Envoy::Http::FilterHeadersStatus::Continue;
```

Comment:
The access log will log request headers mutated by the CER policy. There is not much we can do in this case, but maybe a dd a comment here.

Reply:
Added a comment.